### PR TITLE
Batch events

### PR DESCRIPTION
**Pull request by GabrielRozendo**
Refactoring:
- Unify to get both system info and OS Name once and changed to not null. 
- `InitOptions` in a separated file with more options.
- Normalized `'` and `"`.
- Private fields.
- Not null when possible.
- Handle logs.

Add:
- Storage (as abstract/interface) to persist events for offline and to accumulate before send in batch.
- Check lifecycle, if the app goes to inactive or paused states, send events.
- Send events in batch (customized).
- If an error happens, check to try again later (internet issues or 5xx status code) or discard events (4xx status code).

In this moment, using [Hive](https://pub.dev/packages/hive) but could be implement using anything.

### CODE REVIEW

**Comment by GabrielRozendo** [comment]
Hi @goenning 
Does are PR welcome?


I've been hosting the Aptabase in a small VPS and it is handling huge requests.
I know about PR #6 and as it has a new approach and seems no more active, I'm trying a simpler approach to help (including for my VPS' health 😅).
Did everything thinking in your comments there and approaches used in another SDKs too.

Please check it out and give your feedback.
It is in WIP (work in progress) yet, need to update readme/instructions about them and review some unecessary comments.
Let me know about your thoughts.

Thank you.

**Comment by goenning** [comment]
Hi @GabrielRozendo thanks for the PR, just letting you know that I'll have a look at it this week :)

**Comment by goenning** [comment]
Hi @GabrielRozendo seems pretty good to me!

1. Will it work with flutter on web?
2. Have you tried using it on your app?
3. It's currently in Draft, are you looking to add more things before publishing?

**Comment by GabrielRozendo** [comment]
Hi @goenning 
1: Yes, as Hive and SharedPrefs work for web, it'll be too!
2: Yes, I'm trying in beta right now (2 users) and so far so good. Not released to public yet.
3: I want to check and test more, decide between Hive and SharedPrefs (1), plus, release to my public app and follow how it would be going (2)

It was only a preview to get your firstly impressions and check if it is going good for you.
Good to know!

**Comment by goenning** [comment]
@GabrielRozendo just a heads up that there might be some minor conflicts due to package updates on pubspec.yaml

**Comment by GabrielRozendo** [comment]
@goenning no problem so far and I've found a case.
My server was rebooting (intentionally) and as I was debugging the Flutter app, I validated another scenario.

<details><summary>Logs</summary>
<div>
 
[Aptabase] Checking events, reason: init
[Aptabase] Sent 30 events successfully
[Aptabase] Aptabase initilized!
[Aptabase] TrackEvent failed with status code 502. Response: 
```html
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.22.1</center>
</body>
</html>
```
[Aptabase] Checking events, reason: timer
[Aptabase] Sent 30 events successfully
[Aptabase] TrackEvent failed with status code 502. Response: 
```html
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.22.1</center>
</body>
</html>
```
[Aptabase] Checking events, reason: timer
[Aptabase] TrackEvent Exception: SocketException: Connection refused (OS Error: Connection refused, errno = 61), address = aptabase.myserver.com, port = 51030
[Aptabase] SocketException (SocketException: Connection refused (OS Error: Connection refused, errno = 61), address = aptabase.myserver.com, port = 51030)
[Aptabase]
``` 
           #0      _NativeSocket.startConnect (dart:io-patch/socket_patch.dart:721:35)
           #1      _RawSocket.startConnect (dart:io-patch/socket_patch.dart:1914:26)
           #2      RawSocket.startConnect (dart:io-patch/socket_patch.dart:27:23)
           #3      RawSecureSocket.startConnect (dart:io/secure_socket.dart:299:22)
           #4      SecureSocket.startConnect (dart:io/secure_socket.dart:77:28)
           #5      _ConnectionTarget.connect (dart:_http/http_impl.dart:2483:26)
           #6      _HttpClient._getConnection.connect (dart:_http/http_impl.dart:2926:12)
           #7      _HttpClient._getConnection (dart:_http/http_impl.dart:2931:12)
           #8      _HttpClient._openUrl (dart:_http/http_impl.dart:2786:12)
           #9      _HttpClient.postUrl (dart:_http/http_impl.dart:2633:49)
           #10     Aptabase._send (package:aptabase_flutter/aptabase_flutter.dart:223:35)
           #11     Aptabase._tick (package:aptabase_flutter/aptabase_flutter.dart:144:28)
           <asynchronous suspension>
```
[Aptabase] Checking events, reason: timer
[Aptabase] Sent 30 events successfully
[Aptabase] TrackEvent failed with status code 502. Response:
```html
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.22.1</center>
</body>
</html>
```
[Aptabase] Checking events, reason: timer
[Aptabase] Sent 30 events successfully
[Aptabase] TrackEvent failed with status code 400. Response: Too many events (30) in a single request. Maximum is 25.
[Aptabase] Checking events, reason: timer
[Aptabase] Sent 22 events successfully
[Aptabase] Checking events, reason: timer

</div>
</details> 

---

I'll change this sentence to remove the word successfully. At this point we don't know yet.
`Sent 30 events successfully`

But my real question here is about that 400 response.
The server has a max limit for 25? If so, we should handle this case. I can change that as well.
 
`[Aptabase] TrackEvent failed with status code 400. Response: Too many events (30) in a single request. Maximum is 25.`

**Comment by goenning** [comment]
Yes, there’s a limit of 25 events per request, so the SDK has to create chunks of up to 25 events before sending the request.

**Comment by GabrielRozendo** [comment]
Hey @goenning 
Just updated the PR with `main`

Not in progress anymore.
No problem so far and feel free to review/approve it

File: pubspec.yaml
Comment (cristipufu) [inline pubspec.yaml:1]: We need to update the version of the sdk to 0.4.0:
`version: 0.4.0`

File: lib/aptabase_flutter.dart [hunk 1]
```diff
@@ -1,66 +1,165 @@
...
 ///
-/// Initialize the client with `Aptabase.init(appKey)` and then use `Aptabase.instance.trackEvent(eventName, props)` to record events.
+/// Initialize the client with `Aptabase.init(appKey)` and then
+/// use `Aptabase.instance.trackEvent(eventName, props)` to record events.
 class Aptabase {
-  static const String _sdkVersion = "aptabase_flutter@0.3.0";
-  static const Duration _sessionTimeout = Duration(hours: 1);
+  Aptabase._();
+
+  static const _sdkVersion = "aptabase_flutter@0.3.0";
```
Comment (cristipufu) [inline lib/aptabase_flutter.dart:31]: `static const _sdkVersion = "aptabase_flutter@0.4.0";`

**Review by cristipufu — COMMENTED** [review]
Thanks @GabrielRozendo!

Let's also update the version of the sdk and add a new entry in the [CHANGELOG](https://github.com/aptabase/aptabase_flutter/blob/main/CHANGELOG.md)

**Comment by GabrielRozendo** [comment]
Hi @flutter-painter 

indeed, the `aptabase.init` initialize one storage if you don't provide it, the default one is Hive.

I'm not sure about your specific case but even if you init our storage solution later, as the aptabase init it anyway, I'm not sure if it is a real problem. Maybe you'd have two diff storage solution (one for Aptabase and other for you app).

- Also, you could init the aptabase after init your storage solution, not exactly on main.dart (in my apps I use this approach).
- You can use even other diff than Hive.

In example/main.dart in PR #13 I've included one using SharedPreferences.

**Pull request merged**